components/script/dom/node.rs

```diff
@@ -2129,14 +2287,31 @@ impl NodeMethods for Node {
             NodeTypeId::CharacterData(..) |
             NodeTypeId::Element(..) |
             NodeTypeId::DocumentType |
-            NodeTypeId::DocumentFragment => Some(self.owner_doc()),
+            NodeTypeId::DocumentFragment(_) => Some(self.owner_doc()),
             NodeTypeId::Document(_) => None,
         }
     }
 
     // https://dom.spec.whatwg.org/#dom-node-getrootnode
-    fn GetRootNode(&self) -> DomRoot<Node> {
-        self.inclusive_ancestors().last().unwrap()
+    fn GetRootNode(&self, options: &GetRootNodeOptions) -> DomRoot<Node> {
+        if options.composed {
+            if let Some(shadow_root) = self.containing_shadow_root() {
+                // shadow-including root.
+                return shadow_root.Host().upcast::<Node>().GetRootNode(options);
+            }
+        }
+
+        if let Some(shadow_root) = self.containing_shadow_root() {
+            return DomRoot::from_ref(shadow_root.upcast::<Node>());
```

Comment:
May be worth simplifying this moving the `if options.composed` here, and removing the block above, so:

```rust
if let Some(shadow_root) = self.containing_shadow_root() {
    return if options.composed {
        shadow_root.Host().upcast::<Node>().GetRootNode(options);
    } else {
        DomRoot::from_ref(shadow_root.upcast())
    }
}
```